sst/SimEngCoreWrapper.cc

```diff
@@ -0,0 +1,368 @@
+      // delimiting or escape behavior even with double quotes.
+      // e.g "arg1=1 arg2='"Hi"' arg3=2" will be parsed as
+      // std::vector<std::string>{arg1=1, arg2="Hi", arg3=2}
+      if (currChar == '\'' && escapeSingle) {
+        escapeSingle = 0;
+      }
+      // If a portion of the argument string starts with a double quote (") and
+      // we encounter another double quote, capture the substring enclosed by a
+      // valid set of double quotes into an argument without producing any
+      // delimiting or escape behavior even with single quotes.
+      // e.g "arg1=1 arg2="James' Car" arg3=2" will be parsed as
+      // std::vector<std::string>{arg1=1, arg2=James' Car, arg3=2}
+      else if (currChar == '\"' && escapeDouble) {
+        escapeDouble = 0;
+      } else {
+        str += currChar;
+      }
+    } else {
+      if (currChar == ' ') {
+        if (str != "") {
+          args.push_back(str);
+          str = "";
+        }
+      }
+      // Check for escape character ("), this signals the algorithm to capture
+      // any char inside a set of ("") without producing any delimiting or
+      // escape behavior.
+      else if (currChar == '\"') {
+        escapeDouble = 1;
+        // Check for escape character ('), this signals the algorithm to capture
+        // any char inside a set of ('') without producing any delimiting or
+        // escape behavior.
+      } else if (currChar == '\'') {
+        escapeSingle = 1;
+      } else {
+        str += currChar;
+      }
+    }
+  }
+  if (escapeSingle || escapeDouble) {
+    std::string err;
+    output_.verbose(CALL_INFO, 1, 0, R"(
+           Parsing failed: Invalid format - Please make sure all
+           characters/strings are escaped properly within a set single or 
+           double quotes. To escape quotes use (\\\) instead of (\).\n
+           )");
+    std::cerr << "Error occured at index " << index
+              << " of the argument string - substring: "
+              << "[ " << str << " ]" << std::endl;
+    std::exit(EXIT_FAILURE);
+  }
+  args.push_back(str);
+  return args;
+}
+
+void SimEngCoreWrapper::fabricateSimEngCore() {
+  output_.verbose(CALL_INFO, 1, 0, "Setting up SimEng Core\n");
+  if (simengConfigPath_ != "") {
+#ifdef SIMENG_ENABLE_SST_TESTS
+    if (assembleWithSource_) {
+      output_.verbose(CALL_INFO, 1, 0,
+                      "Assembling source instructions using LLVM\n");
+      Assembler assemble = Assembler(source_);
+      coreInstance_ = std::make_unique<simeng::CoreInstance>(
+          assemble.getAssembledSource(), assemble.getAssembledSourceSize(),
+          simengConfigPath_);
+    } else {
+      coreInstance_ = std::make_unique<simeng::CoreInstance>(
+          simengConfigPath_, executablePath_, executableArgs_);
+    }
+#else
+    coreInstance_ = std::make_unique<simeng::CoreInstance>(
+        simengConfigPath_, executablePath_, executableArgs_);
+#endif
+  } else {
+#ifdef SIMENG_ENABLE_SST_TESTS
```

Comment:
The SST config file should contain all user-specific paths etc., and the contents of an SST config file can be altered through CLI args which is how the test framework works. Only the path to the SST config file is provided in the testing framework. This case handles an extreme case when the path to a SimEng config is not provided within an SST config file.

Each test group specifies an SST config file, and the path to that config file is programmatically determined. Within that SST config file, each test case injects parameters through CLI args. If individual test cases were to contain paths to SimEng configs, then all test cases would have to be edited to specify the correct relative path in different environments.

Also, I'm sure what you mean by "we have no coverage of the code we actually care about", the test suite does target a region of code that the user will use if the test suite was disabled. The test suite gives SimEngCoreWrapper the ability to execute a string of instructions rather than a binary. The test suite does test the integration code by supplying combinations of instructions which target different parts of the integration.  That's why there are test cases with instructions that have misaligned accesses and overlapping address ranges.

Reply:
I can now see that this is just handling the extreme case. If the SST config file is created programatically in the tests however, could the simeng config file be tested and the default injected from there (rather than the coreWrapper). Then there should just be an assert here to catch if that is failing. This will keep as much of the testing framework contained within the testing framework code as possible.

As for the coverage, in the specific case above where a config path is defined, I am referring to the ```CoreInstance``` instantiation. Lines 256 and 261 are essentially the same, however 261 is only run when not testing while 256 is only run when testing. So 261 is not covered.

The main point here is the separation of the testing framework and simulation. As much of the simulation should remain unchanged by the testing framework as possible, so that as much of the simulation is covered. As little as possible should be within the ```ifdef```s. This philosophy being applied throughout, but with this case being solved by moving the ```coreInstance``` instantiation outside of the ```ifdef```

Reply:
I have refactored the code around coreInstance to be more readable and clean. Please take a look. 
Also the sst config isn't generated, the path to it is. The config file is a python file.

Reply:
This is much easier to read, nice work. Noted with the sst config